### ISSUE TITLE
Improve Pylint support

### ISSUE THREAD

**Issue by FabriceSalvaire** [flycheck/flycheck#1802]
Actually `Flycheck Pylintrc` is set to `.pylintrc` by default. But this is not the only way to setup **pylint**, see http://pylint.pycqa.org/en/latest/user_guide/run.html the relevant section is copy-pasted at the end.  Moreover I suspect that in most cases pylint does not found the file due to an incorrect working directory.

If I want to use `setup.cfg`, **flake8** works fine with that, then I have to set this setting in Emacs.  

I think it would be better to tun pylint in the directory that contains a file that suggest it is the source root like `setup.py` or `setup.cfg` and don't use `--rcfile` in the command call.

My hint is to use these standard Python file as an IDE project setup and to locate the working directory.   This working directory must be reported in `show setup` so as to identify weird cases.  

Note pylint works fine with `./src-top/.../src-foo/python_module/foo/.../bar/tata.py` even if `python_module` is the last directory to have a `__init__.py`

Note I don't master elisp programming ...

# Pylint Manual

> Specifying all the options suitable for your setup and coding standards can be tedious, so it is possible to use a configuration file to specify the default values. You can specify a configuration file on the command line using the --rcfile option. Otherwise, Pylint searches for a configuration file in the following order and uses the first one it finds:
> 
>     pylintrc in the current working directory
>     .pylintrc in the current working directory
>     pyproject.toml in the current working directory, providing it has at least one tool.pylint. section.
>     setup.cfg in the current working directory, providing it has at least one pylint. section
>     If the current working directory is in a Python module, Pylint searches up the hierarchy of Python modules until it finds a pylintrc file. This allows you to specify coding standards on a module-by-module basis. Of course, a directory is judged to be a Python module if it contains an __init__.py file.
>     The file named by environment variable PYLINTRC
>     if you have a home directory which isn't /root:
>         .pylintrc in your home directory
>         .config/pylintrc in your home directory
>     /etc/pylintrc

**Comment by cpitclaudel** [comment]
Thanks.  But won't this conflict with https://github.com/flycheck/flycheck/pull/1771 ?

**Comment by FabriceSalvaire** [comment]
I think #1771 solves an issue but is not the right answer.

I would replace that by a top-down walk up to found a pylint setup files or the the file-system root.  Then run pylint with this directory as current working directory.

A weird case would be an isolated Python file with an unrelated setup.cfg file in the directory hierarchy.  But users can solve that by adding a pylint setup file.

Moreover I think we should add a PythonPath setup feature, I don't know if it is easy to parse a file for this purpose.  A workaround is to set the PythonPath before to run Emacs, but personally I work sometimes on more than one project with the same Emacs instance.

I we want to implement IDE features in Emacs then it must work as an IDE and implements projects like Eclipse, Idea ...

**Comment by cpitclaudel** [comment]
> I think #1771 solves an issue but is not the right answer.

I see — but it's the approach used by epylint, right?

> I would replace that by a top-down walk up to found a pylint setup files or the file-system root. Then run pylint with this directory as current working directory.

Won't that break when someone has a .pylintrc in `~`?

> I we want to implement IDE features in Emacs then it must work as an IDE and implements projects like Eclipse, Idea ...

Yes, but that's the job of IDE packages like elpy, I think.

**Comment by FabriceSalvaire** [comment]
I don't know if epylint https://github.com/PyCQA/pylint/commits/master/pylint/epylint.py is still used and was updated.  I used it in the past with flymake.  There is also https://github.com/PyCQA/pylint/blob/master/elisp/pylint.el

The first trick with pylint is the fact it normally expects a Python module like "foo.bar" and eventually the environment variable PYTHONPATH set to something useful, else the algorithm to derive it from the path is voodoo ... I looked the code and had headache ... In some circumstance, the code uses the current working directory to make decision.

Flake8 is also CWD dependent, it must be run >= directory containing the setup file (to perform the top-down walk).  Thus the directory of the input file is fine.

Maybe we should report to pylint bug tracker to have an IDE mode : `pylint --ide absolut_path.py` that do the job, independently of the CWD.  Same could apply to flake8.  The JSON output could also provides the path of the setup files.  pylint --verbose logs it but not flake8.

Is it the job of flycheck to do some magic ?

**There is also another issue with flycheck filename:**

` SqlAlchem…     1   1 info     inval… Module name "flycheck_SqlAlchemyBase" doesn't conform to PascalCase naming style (python-pylint)`

**Comment by cpitclaudel** [comment]
> Is it the job of flycheck to do some magic ?

No, but see https://github.com/flycheck/flycheck/issues/1758; we already opened an issue on the pylint tracker: https://github.com/PyCQA/pylint/issues/3606

> There is also another issue with flycheck filename:

Yes, that will be fixed by https://github.com/flycheck/flycheck/pull/1546

**Comment by OldIMP** [comment]
Would love to know if there's any possibility to use `pyproject.toml` as config instead of .pylintrc.

Thx for the great work!

**Comment by g-braeunlich** [comment]
Does not seem so:
https://github.com/flycheck/flycheck/blob/master/flycheck.el#L10568

**Comment by cpitclaudel** [comment]
I've made one more attempt to solve this at https://github.com/flycheck/flycheck/pull/1771, this time looking for various config files and only falling back to the epylint heuristic if there's no clear project root.  Please give it a go.
